### PR TITLE
GitHub releases

### PR DESCRIPTION
**Pull request by JosephSalisbury**
Add support for creating GitHub releases in the builder release workflow. Fixes https://github.com/giantswarm/builder/issues/63.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -55,7 +55,7 @@ test: $(SOURCE) VERSION .gobuild
 	    --rm \
 	    -v $(shell pwd):/usr/code \
 	    -e GOPATH=/usr/code/.gobuild \
-	    -e GOOS=$(GOOS) \
+	    -e GOOS=linux \
```
Comment (zeisss) [inline Makefile]: ???

File: make/make.go [hunk 2]
```diff
@@ -33,3 +35,24 @@ func BuildProject(clean bool) error {
 
 	return nil
 }
+
+// IsTargetPresent returns true if the given target can be found in
+// the Makefile, false otherwise.
+func IsTargetPresent(makefile string, target string) bool {
+	return strings.Contains(makefile, target)
```
Comment (zeisss) [inline make/make.go]: I guess you want to have a newline before the target and a double colon after it? So I would use a regex here?

File: Makefile (hunk 1 above)
Comment (JosephSalisbury) [inline Makefile]: Running in container, so I've set this because it should always be linux in that case.

File: make/make.go (hunk 2 above)
Comment (JosephSalisbury) [inline make/make.go]: :+1:

File: make/make_test.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+func TestIsTargetPresent(t *testing.T) {
+	var tests = []struct {
+		Makefile string
+		Target   string
+		Expected bool
+	}{
+		{``, ``, true},
+		{``, `all`, false},
+		{`all:
+            test`, `all`, true},
```
Comment (zeisss) [inline make/make_test.go]:
```
all: this is just another target dependency
<tab>echo done!
```

File: make/make.go (hunk 2 above)
Comment (JosephSalisbury) [inline make/make.go]: Not newline, actually - target could be defined on first line of file. Colon is a good shout, though.

Comment (zeisss) [inline make/make.go]: beginning of line then :)

File: make/make_test.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+		Makefile string
+		Target   string
+		Expected bool
+	}{
+		{"", "", false},
+		{"", "all", false},
+		{`all:
+            test`, "all", true},
+		{`# bin-dist
+		all:
```
Comment (zeisss) [inline make/make_test.go]: not sure the backtick notation is good here, as this leads to unnecessary whitespace at the beginning of the test input

Comment (JosephSalisbury) [inline make/make_test.go]: I don't love it either, but I'm not certain on a better method - thoughts?

File: github/github.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
+// Package github provides a client to work with GitHub.
```
Comment (JosephSalisbury) [inline github/github.go]: This is all untested at the moment, fyi

**Comment by JosephSalisbury** [comment]
@giantswarm/cena RFR

File: github/github.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+// Client provides a client to access Github.
+type Client struct {
+	Config Config
+	Client *github.Client
+}
+
+// FindPersonalAccessToken makes a best effort attempt to find a personal access Token
+// to use to authenticate with Github, returning an empty string if one cannot be found.
+func FindPersonalAccessToken() string {
+	return os.Getenv("GITHUB_PERSONAL_ACCESS_TOKEN")
```
Comment (xh3b4sd) [inline github/github.go]: Where does this come from?

File: make/make.go [hunk 7]
```diff
@@ -20,16 +22,54 @@ func MakefileAvailable() bool {
...
 		util.RunCommandExt(exec.Command("make", "clean"), DryRun, Quiet)
 	}
 
-	if _, err := util.RunCommandExt(exec.Command("make"), DryRun, Quiet); err != nil {
+	var command *exec.Cmd
+	if target == "" {
+		command = exec.Command("make")
+	} else {
+		command = exec.Command("make", target)
+	}
```
Comment (xh3b4sd) [inline make/make.go]: Why not simply this? When target is maybe anyway empty, what should happen?

```
command = exec.Command("make", target)
```

File: make/make.go [hunk 8]
```diff
@@ -20,16 +22,54 @@ func MakefileAvailable() bool {
...
+// BuildTarball executes the `bin-dist` target, creating the release tarball.
+func BuildTarball() error {
+	// We don't need to clean first, as we'll have just called make anyway.
+	return runMakeCommand("bin-dist", false)
+}
+
+// IsTargetPresent returns true if the given target can be found in
+// the Makefile, false otherwise.
+func IsTargetPresent(makefile string, target string) bool {
+	return strings.Contains(makefile, target+":")
```
Comment (xh3b4sd) [inline make/make.go]: This should probably use a proper expression. It can only be considered a target in case it starts at the beginning of a line.

File: git/git.go [hunk 9]
```diff
@@ -286,6 +286,19 @@ func (gc *GitClient) Checkout(branch string) error {
...
+func (gc *GitClient) GatherOwnerName() (string, error) {
+	out, err := util.RunCommandExt(exec.Command("git", "config", "--get", "remote.origin.url"), false, gc.Quiet)
+
+	if err != nil {
+		return "", maskAny(err)
+	}
+
+	parts := strings.Split(out, "/")
+
+	return parts[3], nil
```
Comment (zeisss) [inline git/git.go]: thats rather hard focussed on github, isn't it?

File: make/make.go [hunk 10]
```diff
@@ -20,16 +22,54 @@ func MakefileAvailable() bool {
...
+	if !MakefileAvailable() {
+		return false, nil
+	}
+
+	bytes, err := ioutil.ReadFile("Makefile")
+	if err != nil {
+		return false, errgo.Mask(err)
+	}
+
+	return IsTargetPresent(string(bytes), "bin-dist"), nil
```
Comment (xh3b4sd) [inline make/make.go]: So when we consider some special directory like `bin-dist` then `builder` is going to be even more specialized in the giantswarm direction.

File: task/release.go [hunk 11]
```diff
@@ -108,6 +119,62 @@ func (rt *ReleaseTask) Run() error {
 		}
 	}
 
+	personalAccessToken := github.FindPersonalAccessToken()
```
Comment (zeisss) [inline task/release.go]: can we do this check earlier so we get an error BEFORE we actually do the release?

File: task/release.go [hunk 12]
```diff
@@ -108,6 +119,62 @@ func (rt *ReleaseTask) Run() error {
...
+		fmt.Println("Could not find token for GitHub, will not create GitHub release")
+	}
+
+	hasBinDistTarget, err := make.HasBinDistTarget()
+	if err != nil {
+		return errgo.Mask(err)
+	}
+
+	if personalAccessToken != "" && hasBinDistTarget {
+		fmt.Println("bin-dist target found")
```
Comment (zeisss) [inline task/release.go]: I like. Can we also have an `else` with a verbose message? 

> No 'bin-dist'  target found in Makefile.

and if the token is empty a warning maybe?

> Found 'bin-dist'  target in Makefile, but no github token was found. 

I guess the last case should probably be an error?

File: task/release.go [hunk 13]
```diff
@@ -108,6 +119,62 @@ func (rt *ReleaseTask) Run() error {
...
+		}
+
+		// Find the tarball to upload
+		var tarballFilename string
+		fileInfos, err := ioutil.ReadDir("bin-dist")
+		if err != nil {
+			return errgo.Mask(err)
+		}
+		for _, fileInfo := range fileInfos {
+			if strings.HasSuffix(fileInfo.Name(), "tar.gz") {
```
Comment (zeisss) [inline task/release.go]: `.tar.gz`

File: github/github.go (hunk 6 above)
Comment (JosephSalisbury) [inline github/github.go]: Invented - if there is a standard for this, I couldn't find it.

File: git/git.go (hunk 9 above)
Comment (JosephSalisbury) [inline git/git.go]: Yes, but the notion of an `owner` is itself fairly github focused, so uncertain on a better solution

Comment (zeisss) [inline git/git.go]: Absolutely. But lets add an assert or sthg that `parts[2]` is `"github.com"` or sthg.

File: make/make.go (hunk 7 above)
Comment (JosephSalisbury) [inline make/make.go]: +1

File: task/release.go (hunk 13 above)
Comment (JosephSalisbury) [inline task/release.go]: +1

File: make/make.go (hunk 8 above)
Comment (JosephSalisbury) [inline make/make.go]: +1

File: make/make.go (hunk 10 above)
Comment (JosephSalisbury) [inline make/make.go]: I reckon `builder` is always going to be somewhat opinionated to our workflow. We don't want to be building a generic build tool.

File: task/release.go (hunk 11 above)
Comment (JosephSalisbury) [inline task/release.go]: +1

File: task/release.go (hunk 12 above)
Comment (JosephSalisbury) [inline task/release.go]: +1

File: make/make.go (hunk 7 above)
Comment (JosephSalisbury) [inline make/make.go]: Ah, this is why I went this way:
```make: *** empty string invalid as file name.  Stop.```

Comment (xh3b4sd) [inline make/make.go]: Ok, I guess there is some escaping magic that turns the empty string into a separate argument. What about this? 

```
command = exec.Command("make", []string{target}...)
```

**Comment by JosephSalisbury** [comment]
@zeisss @xh3b4sd :fishing_pole_and_fish:

File: git/git.go [hunk 14]
```diff
@@ -286,6 +286,23 @@ func (gc *GitClient) Checkout(branch string) error {
 	return maskAny(err)
 }
 
+// GatherOwnerName returns the name of the repository's owner.
+func (gc *GitClient) GatherOwnerName() (string, error) {
```
Comment (xh3b4sd) [inline git/git.go]: I would like to see tests for this.

**Comment by xh3b4sd** [comment]
May @denderello also has an opinion on that.

File: git/git.go (hunk 14 above)
Comment (JosephSalisbury) [inline git/git.go]: Yeah, I like that, I shall do so.

File: github/github.go (hunk 6 above)
Comment (denderello) [inline github/github.go]: The [`hub`](https://github.com/github/hub) command which is developed by the github team uses a config file in `~/.config/hub` for this. It holds the following parameters:
```
github.com:
- user: <username>
  oauth_token: <token>
  protocol: <scheme>
```

How about doing something like this for builder too? I can think of a file in the users home directory that holds configuration for things like github tokens, etc. and local config files in the projects that configure docker registry endpoints, etc.

Comment (denderello) [inline github/github.go]: Oh and https://github.com/github/hub/blob/5b3aa70614fee34ae6fd6445ea881d4d1632ce79/github/config.go is a nice resource to see how they implemented things. If the file is not present the users get's prompted to input the credentials which will then be saved in the config file. very convenient imho.

File: make/make.go (hunk 7 above)
Comment (denderello) [inline make/make.go]: Yes, this empty strings will passed to `execve(2)`'s `argv` parameter since Go is pretty strict about this. Your shells will usually just trim the command line you just passed to them so that stuff like this will not happen. We have to take care about this ourselves.

File: task/release.go [hunk 15]
```diff
@@ -108,6 +138,50 @@ func (rt *ReleaseTask) Run() error {
...
+			PersonalAccessToken: personalAccessToken,
+		}
+		githubClient := github.NewClient(config)
+
+		fmt.Println("Building release tarball")
+		if err := make.BuildTarball(); err != nil {
+			return errgo.Mask(err)
+		}
+
+		// Find the tarball to upload
```
Comment (denderello) [inline task/release.go]: How about putting this into a separate package? To me this is logic that is not about the orchestration of a task but about finding specific files in an arbitrary directory.

File: make/make.go [hunk 16]
```diff
@@ -20,16 +23,55 @@ func MakefileAvailable() bool {
...
 
 	return nil
 }
+
+func BuildProject(clean bool) error {
+	return runMakeCommand("", clean)
+}
+
+// BuildTarball executes the `bin-dist` target, creating the release tarball.
+func BuildTarball() error {
```
Comment (denderello) [inline make/make.go]: I would not call this `BuildTarball` since the `bin-dist` target can produce any kind of artifact. The `make` package imho should not care about knowing that it provides a tarball. It just calls a target that produces artifacts. What others are doing with those is out of context of this package.

File: make/make.go (hunk 10 above)
Comment (denderello) [inline make/make.go]: I copy that. We are seeing more and more that `builder` needs to have a workflow like `maven` with given steps in it. I personally like approaches that those steps only get triggered if certain circumstances are true.

**Comment by denderello** [comment]
I like the approach in general. We have to think about the workflow of this a bit I would say. Aside from this documentation for the new feature is missing.

File: github/github.go (hunk 6 above)
Comment (JosephSalisbury) [inline github/github.go]: I like this, but I'd like to do it outside of this PR.

File: Makefile (hunk 1 above)
Comment (zeisss) [inline Makefile]: I would like to remove this before merging. afaik it already defaults to linux if you leave this empty.

File: README.md [hunk 17]
```diff
@@ -98,6 +98,16 @@ The urls of the private and public registry are fixed, but you can override them
 using either the `--docker-private-registry` or the `--docker-public-registry`
 flags.
 
+### GitHub Releases
+
+To create GitHub releases, `builder` will search for a target in your `Makefile`
+called `bin-dist`. If this target is present, and the environment variable
+`GITHUB_PERSONAL_ACCESS_TOKEN` is non-empty, `builder` will build the `bin-dist`
+target, create a GitHub release for the tag being released, and upload a
+file found in a directory named `bin-dist` ending with `.tar.gz` to the release.
```
Comment (zeisss) [inline README.md]: "a file" or "all files"? I thought we want to support multiple?

File: github/github.go (hunk 6 above)
Comment (denderello) [inline github/github.go]: Fine for me to keep the PR smaller

File: util/file.go [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+// and a pointer to the opened file for a file that matches the predicate.
+func FindFileInDirectory(directory string, predicate func(string) bool) (string, *os.File, error) {
+	var filename string
+	fileInfos, err := ioutil.ReadDir(directory)
+	if err != nil {
+		return "", nil, errgo.Mask(err)
+	}
+	for _, fileInfo := range fileInfos {
+		if predicate(fileInfo.Name()) {
+			filename = fileInfo.Name()
```
Comment (zeisss) [inline util/file.go]: you can break here

File: task/release.go [hunk 19]
```diff
@@ -52,6 +55,25 @@ func (rt *ReleaseTask) Run() error {
...
 	}
 
+	hasBinDistTarget, err := make.HasBinDistTarget()
+	if err != nil {
+		return errgo.Mask(err)
+	}
+	if hasBinDistTarget {
+		fmt.Println("bin-dist target found")
+	} else {
+		fmt.Println("bin-dist target not found in makefile")
```
Comment (zeisss) [inline task/release.go]: lets write Makefile here (uppercase M), as this is actually the file we look at

**Comment by zeisss** [comment]
I am not sure about the fact we only support uploading ONE file to a release. Do we have projects where we want multiple binaries? For now `builder` comes to my mind, maybe `formicactl` or `releaseit`. Other tools are mostly usable only on linux, so thats probably not a problem.

File: Makefile (hunk 1 above)
Comment (JosephSalisbury) [inline Makefile]: Will do

File: util/file.go (hunk 18 above)
Comment (JosephSalisbury) [inline util/file.go]: Will do

File: task/release.go (hunk 19 above)
Comment (JosephSalisbury) [inline task/release.go]: Will do

**Comment by JosephSalisbury** [comment]
That's a good point, I hadn't thought of that. It sounds like we come back to the "how to specify which os to build for" problem, though.

**Comment by zeisss** [comment]
@JosephSalisbury Why? If we expect `make bin-dist` to prepare all tarballs inside the `bin-dist/` folder, we just upload everything that is in there. If the project wants to have a linux + darwin release, it should build the two tarballs themself.

**Comment by zeisss** [comment]
Afaik we can add this next, doesn't block this PR. LGTM

**Comment by JosephSalisbury** [comment]
@zeisss sorry, I get what you mean now - we need to add support for _multiple_ tarballs, with it. I'll open an issue for it.

File: README.md (hunk 17 above)
Comment (JosephSalisbury) [inline README.md]: As discussed later, "a file" for now, "all files" in future.
